Fix Zig 0.15.2 linking on macOS 26 (Xcode 26.4).

macOS 26 SDK dropped `arm64-macos` from TBD stub files, only listing
`arm64e-macos`. This causes Zig's MachO linker to fail to resolve any
libSystem symbols when targeting aarch64-macos.

Backported from https://codeberg.org/ziglang/zig/pulls/31673

diff --git a/src/link/MachO/Dylib.zig b/src/link/MachO/Dylib.zig
index 730ae10cc..abc5de02 100644
--- a/src/link/MachO/Dylib.zig
+++ b/src/link/MachO/Dylib.zig
@@ -730,29 +730,35 @@ pub const TargetMatcher = struct {
             .cpu_arch = cpu_arch,
             .platform = platform,
         };
-        const apple_string = try targetToAppleString(allocator, cpu_arch, platform);
-        try self.target_strings.append(allocator, apple_string);
+        try self.addTargetStrings(cpuArchToAppleString(cpu_arch));
+        // In Xcode 26.4, Apple unified their TBD files from having separate `arm64-macos` and `arm64e-macos`
+        // entries to having just the latter, presumably because the symbol lists are identical anyway. It
+        // sure would have been nice if they settled on the former as the unified name so as not to break the
+        // world, but evidently we can't have nice things.
+        if (cpu_arch == .aarch64) try self.addTargetStrings("arm64e");

-        switch (platform) {
-            .IOSSIMULATOR, .TVOSSIMULATOR, .WATCHOSSIMULATOR, .VISIONOSSIMULATOR => {
-                // For Apple simulator targets, linking gets tricky as we need to link against the simulator
-                // hosts dylibs too.
-                const host_target = try targetToAppleString(allocator, cpu_arch, .MACOS);
-                try self.target_strings.append(allocator, host_target);
-            },
-            .MACOS => {
-                // Turns out that around 10.13/10.14 macOS release version, Apple changed the target tags in
-                // tbd files from `macosx` to `macos`. In order to be compliant and therefore actually support
-                // linking on older platforms against `libSystem.tbd`, we add `<cpu_arch>-macosx` to target_strings.
-                const fallback_target = try std.fmt.allocPrint(allocator, "{s}-macosx", .{
-                    cpuArchToAppleString(cpu_arch),
-                });
-                try self.target_strings.append(allocator, fallback_target);
-            },
+        return self;
+    }
+
+    fn addTargetStrings(self: *TargetMatcher, arch: []const u8) !void {
+        try self.target_strings.append(self.allocator, try std.fmt.allocPrint(
+            self.allocator,
+            "{s}-{s}",
+            .{ arch, platformToAppleString(self.platform) },
+        ));
+
+        switch (self.platform) {
+            .MACCATALYST => {
+                // Mac Catalyst is allowed to link macOS libraries in a TBD because Apple were apparently too lazy
+                // to add the proper target strings despite doing so in other places in the format???
+                try self.target_strings.append(self.allocator, try std.fmt.allocPrint(self.allocator, "{s}-macos", .{arch}));
+            },
+            .IOSSIMULATOR, .TVOSSIMULATOR, .WATCHOSSIMULATOR, .VISIONOSSIMULATOR => {
+                // For Apple simulator targets, we need to link against the simulator host's libraries too.
+                try self.target_strings.append(self.allocator, try std.fmt.allocPrint(self.allocator, "{s}-macos", .{arch}));
+            },
             else => {},
         }
-
-        return self;
     }

     pub fn deinit(self: *TargetMatcher) void {
@@ -762,7 +768,7 @@ pub const TargetMatcher = struct {
         self.target_strings.deinit(self.allocator);
     }

-    inline fn cpuArchToAppleString(cpu_arch: std.Target.Cpu.Arch) []const u8 {
+    fn cpuArchToAppleString(cpu_arch: std.Target.Cpu.Arch) []const u8 {
         return switch (cpu_arch) {
             .aarch64 => "arm64",
             .x86_64 => "x86_64",
@@ -770,9 +776,8 @@ pub const TargetMatcher = struct {
         };
     }

-    pub fn targetToAppleString(allocator: Allocator, cpu_arch: std.Target.Cpu.Arch, platform: macho.PLATFORM) ![]const u8 {
-        const arch = cpuArchToAppleString(cpu_arch);
-        const plat = switch (platform) {
+    fn platformToAppleString(platform: macho.PLATFORM) []const u8 {
+        return switch (platform) {
             .MACOS => "macos",
             .IOS => "ios",
             .TVOS => "tvos",
@@ -787,7 +792,6 @@ pub const TargetMatcher = struct {
             .DRIVERKIT => "driverkit",
             else => unreachable,
         };
-        return std.fmt.allocPrint(allocator, "{s}-{s}", .{ arch, plat });
     }

     fn hasValue(stack: []const []const u8, needle: []const u8) bool {
